### ISSUE TITLE
Impossible to provide -DCRYPTOPP_DATA_DIR in CXXFLAGS

### ISSUE THREAD

**Issue by mouse07410** [weidai11/cryptopp#440]
The goal is to be able to run `cryptest.exe` (including `cryptest.exe v` etc.) after the installation has completed, possibly when the `cryptopp` source directory has been removed.

Any attempt to supply `CRYPTOPP_DATA_DIR` flag seems to fail:
```
$ echo $CXXFLAGS
-maes -mpclmul -mrdrnd -msse2 -mssse3 -msse4.1 -msse4.2 -std=c++11 -Os -Ofast -DCRYPTOPP_DATA_DIR="/opt/local/share/cryptopp/"
. . . . .
clang++ -maes -mpclmul -mrdrnd -msse2 -mssse3 -msse4.1 -msse4.2 -std=c++11 -Os -Ofast -DCRYPTOPP_DATA_DIR="/opt/local/share/cryptopp/" -fPIC -march=native -pipe -c validat0.cpp
test.cpp:393:21: error: expected expression
                        FileSource usage(CRYPTOPP_DATA_DIR "TestData/usage.dat", true, new F...
                                         ^
<command line>:1:27: note: expanded from here
#define CRYPTOPP_DATA_DIR /opt/local/share/cryptopp/
                          ^
test.cpp:393:21: error: use of undeclared identifier 'opt'
<command line>:1:28: note: expanded from here
#define CRYPTOPP_DATA_DIR /opt/local/share/cryptopp/
                           ^
test.cpp:393:21: error: use of undeclared identifier 'local'
<command line>:1:32: note: expanded from here
#define CRYPTOPP_DATA_DIR /opt/local/share/cryptopp/
                               ^
test.cpp:393:21: error: use of undeclared identifier 'share'
<command line>:1:38: note: expanded from here
#define CRYPTOPP_DATA_DIR /opt/local/share/cryptopp/
                                     ^
test.cpp:393:21: error: use of undeclared identifier 'cryptopp'
<command line>:1:44: note: expanded from here
#define CRYPTOPP_DATA_DIR /opt/local/share/cryptopp/
                                           ^
5 errors generated.
```

I don't know what the following code in `GNUmakefile` does
```
# Only use CRYPTOPP_DATA_DIR if its not set in CXXFLAGS
ifeq ($(findstring -DCRYPTOPP_DATA_DIR, $(strip $(CXXFLAGS))),)
ifneq ($(strip $(CRYPTOPP_DATA_DIR)),)
validat%.o : validat%.cpp
        $(CXX) $(strip $(CXXFLAGS)) -DCRYPTOPP_DATA_DIR=\"$(CRYPTOPP_DATA_DIR)\" -c $<
bench%.o : bench%.cpp
        $(CXX) $(strip $(CXXFLAGS)) -DCRYPTOPP_DATA_DIR=\"$(CRYPTOPP_DATA_DIR)\" -c $<
datatest.o : datatest.cpp
        $(CXX) $(strip $(CXXFLAGS)) -DCRYPTOPP_DATA_DIR=\"$(CRYPTOPP_DATA_DIR)\" -c $<
test.o : test.cpp
        $(CXX) $(strip $(CXXFLAGS)) -DCRYPTOPP_DATA_DIR=\"$(CRYPTOPP_DATA_DIR)\" -c $<
endif
endif
```
but based on my experience on MacOS it is either not enough, or not correct.

I think the options are:
- copy/install test vectors and data to the `CRYPTOPP_DATA_DIR` before invoking `./cryptest.exe` in the GNUmakefile (target `test`), aka perform partial installation before running the test;
- in the test-modules like `validat.cpp` check for the existence of `CRYPTOPP_DATA_DIR`, and if it is not found - use the current directory.

I lean towards the second option as the more robust one.

**Comment by noloader** [comment]
>     #define CRYPTOPP_DATA_DIR /opt/local/share/cryptopp/

According to [DataDir](https://www.cryptopp.com/wiki/DataDir), the `CRYPTOPP_DATA_DIR` macro is a string and uses simple concatenation. So this should work for you:

```
#define CRYPTOPP_DATA_DIR "/opt/local/share/cryptopp/"
```

-----

>         $(CXX) $(strip $(CXXFLAGS)) -DCRYPTOPP_DATA_DIR=\"$(CRYPTOPP_DATA_DIR)\" -c $<

What happens above is, the makefile variable `$(CRYPTOPP_DATA_DIR)` is substituted with an environmental variable (if its set). The `\"` escapes the quote, so you are left with soething like:

```
$(CXX) $(CXXFLAGS) -DCRYPTOPP_DATA_DIR="/opt/local/share/cryptopp/" -c $<
```

I think you would invoke it like:

```
CRYPTOPP_DATA_DIR="/opt/local/share/cryptopp/" make -j 4
```

**Comment by mouse07410** [comment]
> So this should work for you:
> `#define CRYPTOPP_DATA_DIR "/opt/local/share/cryptopp/"`

Not quite. The above works - but only after the test vectors and such *have been installed* to that directory. If you choose have successful `make test` as a prerequisite to `sudo make install`, the above fails.

> I think you would invoke it like:
> `CRYPTOPP_DATA_DIR="/opt/local/share/cryptopp/" make -j 4`

I did not think of this - looks nice, but it should suffer from the same problem I described in this post: if you specify `CRYPTOPP_DATA_DIR`, it must be populated *before* you do `make test`. Which breaks the idea of a fresh installation that should also work after the source tree has been removed.

The risks of not deleting the previous test vectors are obvious, so that's not an attractive path. 

Also, the problem is not with the `make` stage - that one invariably succeeds. The problem is with the `make test` stage, which invariably fails: if you specify `CRYPTOPP_DATA_DIR` then for a fresh installation it is empty, and if you don't - the built `cryptest.exe` cannot run anywhere except the source directory tree.

**Comment by noloader** [comment]
> Not quite. The above works - but only after the test vectors and such have been installed to that directory. If you choose have successful make test as a prerequisite to sudo make install, the above fails.

Yeah, the original patch looked in `PWD` and `CRYPTOPP_DATA_DIR` for the data files and test vectors. When we solicited feedback from the distros, they did not want paths searched for the data files and test vectors. They felt it could be a vulnerability by getting the `cryptest.exe` program to consume arbitrary and untrusted input.

The compromise was the double build.

**Comment by mouse07410** [comment]
> The compromise was the double build.

So one either has to install and only then find out whether the code works or not, or build and test in the current directory then install where it's supposed to be and then run the tests again from outside the source directory...

OK...

**Comment by noloader** [comment]
@mouse07410,

Sorry to loop back to this... What if we did this here...

1. If nothing special is done, and only `CRYPTOPP_DATA_DIR` is supplied, then do existing behavior (the double build). This follows the principle of least surprise, and avoids the distro concerns.

2. If `CRYPTOPP_DATA_DIR` is supplied ***and*** `CRYPTOPP_DATA_DIR_PWD` is defined, then check both the present working directory and the data directory. This requires special action for people who desire the behavior.

What do you think?

**Comment by mouse07410** [comment]
It probably beats the need to run a double build.

I am for implementing this feature. Especially if in the docs we make clear what the risks and benefits of this feature are.

**Comment by noloader** [comment]
@mouse07410,

If we adopt strategy (2), then we need to change:

```
pass=RunTestDataFile(CRYPTOPP_DATA_DIR "TestVectors/keccak.txt") && pass;
```

To something like:

```
pass=RunTestDataFile(GetDataDirectory("TestVectors/keccak.txt")) && pass;
```

Are you OK with it?

**Comment by mouse07410** [comment]
> `AddDataDirectory()` will have to be a non-inlined function. I guess that begs adding `datadir.c` (or stuffing it in an existing test file).

Since this is not a part of the "main" library - i.e., size and performance optimizations are of no concern - why not?

> Are you OK with it?

Sure. I see no real downside here. Am I missing anything...?
